File: 20-actors/niyaku/methods/test_terminal_cycle.py
"""Tests for terminal_cycle — end-to-end vessel-discharge orchestration."""

import pytest

from agv_transfer import Agv
from crane_dynamics import GantryCrane
from stow_plan import Container
from terminal_cycle import DischargeReport, YardLayout, simulate_discharge


def _boxes(n, port="JPYOK"):
    return [Container(f"B{i}", 20.0 - i, port) for i in range(n)]


def test_basic_discharge_runs_all_boxes():
    r = simulate_discharge(_boxes(6), ["JPYOK"], "JPYOK", bays=2, rows=2, tiers=3)
    assert isinstance(r, DischargeReport)
    assert r.moves == 6
    assert len(r.records) == 6
    assert r.discharge_time_s > 0
    assert 10 < r.moves_per_hour() < 200       # sane STS productivity band
    assert all(rec.agv_id for rec in r.records)  # every box got an AGV


def test_only_target_port_discharged():
    boxes = (
        _boxes(3, "JPYOK")
        + [Container(f"R{i}", 15.0, "NLRTM") for i in range(3)]
    )
    r = simulate_discharge(boxes, ["JPYOK", "NLRTM"], "JPYOK",
                           bays=3, rows=2, tiers=3)
    assert r.moves == 3
    assert all(rec.box_id.startswith("B") for rec in r.records)


def test_discharge_is_max_of_crane_and_agv():
    r = simulate_discharge(_boxes(4), ["JPYOK"], "JPYOK", bays=2, rows=2, tiers=2)
    assert r.discharge_time_s == pytest.approx(max(r.crane_timeline_s, r.agv_makespan_s))


def test_more_agvs_do_not_raise_crane_bound_time():
    """The single STS crane is the bottleneck; adding AGVs cannot speed it past
    the serial crane timeline."""
    common = dict(rotation=["JPYOK"], discharge_port="JPYOK", bays=2, rows=2, tiers=3)
    r2 = simulate_discharge(_boxes(6), agv_ids=["A1", "A2"], **common)
    r5 = simulate_discharge(_boxes(6), agv_ids=["A1", "A2", "A3", "A4", "A5"], **common)
    assert r2.crane_timeline_s == pytest.approx(r5.crane_timeline_s)
    assert r5.agv_makespan_s <= r2.agv_makespan_s
    # crane-bound: overall time equals the crane timeline in both
    assert r5.discharge_time_s == pytest.approx(r5.crane_timeline_s)


def test_empty_port_zero_productivity():
    r = simulate_discharge(_boxes(3, "JPYOK"), ["JPYOK", "SGSIN"], "SGSIN",
                           bays=2, rows=2, tiers=2)
    assert r.moves == 0
    assert r.discharge_time_s == 0.0
    assert r.moves_per_hour() == 0.0


def test_accepts_prebuilt_plan():
    from stow_plan import build_stow_plan
    boxes = _boxes(4)
    plan = build_stow_plan(boxes, ["JPYOK"], bays=2, rows=2, tiers=2)
    r = simulate_discharge(boxes, ["JPYOK"], "JPYOK", bays=2, rows=2, tiers=2, plan=plan)
    assert r.moves == 4


def test_custom_crane_yard_agv():
    r = simulate_discharge(
        _boxes(3), ["JPYOK"], "JPYOK", bays=2, rows=2, tiers=2,
        crane=GantryCrane(cable_length=20.0), agv=Agv(v_max=4.0),
        yard=YardLayout(apron_to_yard_m=200.0),
    )
    assert r.moves == 3
    assert r.max_residual_sway_m >= 0.0


def test_isaac_path_runs_or_falls_back():
    """use_isaac=True must produce a valid report whether or not the Isaac
    surface is importable (it falls back to the analytic model)."""
    r = simulate_discharge(_boxes(3), ["JPYOK"], "JPYOK",
                           bays=2, rows=2, tiers=2, use_isaac=True)
    assert r.moves == 3
    assert r.discharge_time_s > 0

File: 20-actors/niyaku/methods/test_terminal_cycle.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_terminal_cycle.py.
(ns niyaku.methods.test-terminal-cycle
  "Tests for terminal_cycle — end-to-end vessel-discharge orchestration.

  Run:  bb --classpath 20-actors 20-actors/niyaku/methods/test_terminal_cycle.clj"
  (:require [niyaku.methods.terminal-cycle :as tc]
            [niyaku.methods.crane-dynamics :as cd]
            [niyaku.methods.agv-transfer :as agv]
            [niyaku.methods.stow-plan :as st]
            [clojure.test :refer [deftest is run-tests]]))

(defn- approx [a b] (<= (Math/abs (double (- a b))) 1e-6))
(defn- boxes
  ([n] (boxes n "JPYOK"))
  ([n port] (map-indexed (fn [i _] (st/container (str "B" i) (- 20.0 i) port)) (range n))))

(deftest basic-discharge-runs-all-boxes
  (let [r (tc/simulate-discharge (boxes 6) ["JPYOK"] "JPYOK" 2 2 3)]
    (is (= (:moves r) 6))
    (is (= (count (:records r)) 6))
    (is (> (:discharge-time-s r) 0))
    (is (< 10 (tc/moves-per-hour r) 200))
    (is (every? #(seq (:agv-id %)) (:records r)))))   ; every box got an AGV

(deftest only-target-port-discharged
  (let [bs (concat (boxes 3 "JPYOK")
                   (map-indexed (fn [i _] (st/container (str "R" i) 15.0 "NLRTM")) (range 3)))
        r (tc/simulate-discharge bs ["JPYOK" "NLRTM"] "JPYOK" 3 2 3)]
    (is (= (:moves r) 3))
    (is (every? #(clojure.string/starts-with? (:box-id %) "B") (:records r)))))

(deftest discharge-is-max-of-crane-and-agv
  (let [r (tc/simulate-discharge (boxes 4) ["JPYOK"] "JPYOK" 2 2 2)]
    (is (approx (:discharge-time-s r) (max (:crane-timeline-s r) (:agv-makespan-s r))))))

(deftest more-agvs-do-not-raise-crane-bound-time
  (let [r2 (tc/simulate-discharge (boxes 6) ["JPYOK"] "JPYOK" 2 2 3 :agv-ids ["A1" "A2"])
        r5 (tc/simulate-discharge (boxes 6) ["JPYOK"] "JPYOK" 2 2 3
                                  :agv-ids ["A1" "A2" "A3" "A4" "A5"])]
    (is (approx (:crane-timeline-s r2) (:crane-timeline-s r5)))
    (is (<= (:agv-makespan-s r5) (:agv-makespan-s r2)))
    (is (approx (:discharge-time-s r5) (:crane-timeline-s r5)))))   ; crane-bound

(deftest empty-port-zero-productivity
  (let [r (tc/simulate-discharge (boxes 3 "JPYOK") ["JPYOK" "SGSIN"] "SGSIN" 2 2 2)]
    (is (= (:moves r) 0))
    (is (= (:discharge-time-s r) 0.0))
    (is (= (tc/moves-per-hour r) 0.0))))

(deftest accepts-prebuilt-plan
  (let [bs (boxes 4)
        plan (st/build-stow-plan bs ["JPYOK"] 2 2 2)
        r (tc/simulate-discharge bs ["JPYOK"] "JPYOK" 2 2 2 :plan plan)]
    (is (= (:moves r) 4))))

(deftest custom-crane-yard-agv
  (let [r (tc/simulate-discharge (boxes 3) ["JPYOK"] "JPYOK" 2 2 2
                                 :crane (cd/make-crane :cable-length 20.0)
                                 :agv (agv/make-agv :v-max 4.0)
                                 :yard (tc/yard-layout :apron-to-yard-m 200.0))]
    (is (= (:moves r) 3))
    (is (>= (:max-residual-sway-m r) 0.0))))

(deftest isaac-path-runs-or-falls-back
  ;; use-isaac=true must produce a valid report (clj port falls back to the analytic model)
  (let [r (tc/simulate-discharge (boxes 3) ["JPYOK"] "JPYOK" 2 2 2 :use-isaac true)]
    (is (= (:moves r) 3))
    (is (> (:discharge-time-s r) 0))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'niyaku.methods.test-terminal-cycle)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
